### PR TITLE
python311Packages.finetuning-scheduler: init at 2.3.3

### PR DESCRIPTION
**Pull request by bcdarwin**
## Description of changes

Add `finetuning-scheduler`, a package for generating Pytorch/Lightning learning rate schedules.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by bcdarwin** [comment]
Marking draft until pytorch-lightning gets bumped.

**Comment by bcdarwin** [comment]
@ofborg build python311Packages.finetuning-scheduler

File: pkgs/development/python-modules/finetuning-scheduler/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    repo = "finetuning-scheduler";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-Nci+NcfcaZV3NH3MTCyr8PwPdDykJTdq8+CmtngWdiM=";
+  };
+
+  build-system = [
+    setuptools
+  ];
+
+  propagatedBuildInputs = [
```
Comment (natsukium) [inline pkgs/development/python-modules/finetuning-scheduler/default.nix:29]:
```suggestion
  dependencies = [
```

**Review by natsukium — COMMENTED** [review]
Could you format it with `nixfmt-rfc-style`?

**Comment by natsukium** [comment]
regarding aarch64-linux
```
E               assert (True, True, ...678, 0.688296) == (True, True, ...678, 0.688297)
E                 
E                 At index 5 diff: 0.688296 != 0.688297
```
```
       > =========================== short test summary info ============================
       > FAILED tests/test_finetuning_scheduler_callback.py::test_fts_frozen_bn_track_running_stats[frozen_bn_track] - assert (True, False,...678, 0.688296) == (True, False,...678, 0.688297)
       > FAILED tests/test_finetuning_scheduler_callback.py::test_fts_frozen_bn_track_running_stats[no_frozen_bn_track] - assert (True, True, ...678, 0.688296) == (True, True, ...678, 0.688297)
       > ===== 2 failed, 118 passed, 43 skipped, 513 warnings in 348.62s (0:05:48) ======
```

It looks like these tests should be disabled.

File: pkgs/development/python-modules/finetuning-scheduler/default.nix (hunk 1 above)
Comment (bcdarwin) [reply pkgs/development/python-modules/finetuning-scheduler/default.nix]: Done, sorry.

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 318695` at 0b308faa run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.finetuning-scheduler</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/natsukium/300eca2d2e4558e7e6017779648ce0bd">python311Packages.finetuning-scheduler</a></li>
  </ul>
</details>

**Comment by bcdarwin** [comment]
Marked broken on Darwin due to `torch.distributed` issues.

File: pkgs/development/python-modules/finetuning-scheduler/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  pytestFlagsArray = [ "tests" ];
+  disabledTests = lib.optionals (stdenv.isAarch64 && stdenv.isLinux) [
+    # slightly exceeds numerical tolerance on aarch64-linux:
+    "test_fts_frozen_bn_track_running_stats"
+  ];
+
+  pythonImportsCheck = [ "finetuning_scheduler" ];
+
+  meta = {
+    description = "A PyTorch Lightning extension for foundation model experimentation with flexible fine-tuning schedules";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/finetuning-scheduler/default.nix:47]:
```suggestion
    description = "PyTorch Lightning extension for foundation model experimentation with flexible fine-tuning schedules";
```

**Review by natsukium — APPROVED** [review]
LGTM, thanks.
